pkgs/README: document deprecation-before-removal guidelines (#520061)
NixOS/nixpkgs: pkgs/README.md

## Deprecating/removing packages

Before removing a package, try to find a new maintainer or fix smaller issues first.

If your package is widely used (by other packages, NixOS modules, or end-user configs), give people more warnings before you remove it.

### Deprecation before removal

> Evaluation-time warnings are the only channel that reliably reaches every user.

Reviewer: FWIW, my impression is that they don’t actually reliably reach everyone (e.g. automatic updates), though that’s a harder problem to solve.

Reviewer 2: I guess we could reconsider this as a wording nit. Something like:
```suggestion
Evaluation-time warnings are the most likely channel to reach users.
```

I'm not sure it's worth explicitly caveating that even eval warnings can be missed, especially when automation is involved.

> You should include removals/deprecations into the release notes.

Reviewer: I believe that we should actually put far less into the release notes. Anything that the computer can tell you by simply evaluating your configuration is better left out of the release notes unless it’s extremely high‐impact: a programmatic error message is more targeted than a release note could ever be, and the current release notes are so bogged down by things most users will never care about that it discourages reading them for the things that we *can’t* notify people about programmatically. Putting every package drop in them would be a regression.

Author: Ok. Agreed i'll remove.

Reviewer 2: On that basis, is it worth including something along the lines of:

```suggestion
Only significant removals/deprecations should be mentioned in the release notes.
```

?

It seems many of us are "following precedent" and including trivial removals in the release notes, it'd be helpful to have a sentence to point at when suggesting such release notes be omitted.

**When a replacement exists**, use `warnAlias` in `pkgs/top-level/aliases.nix` to use the replacement while emitting an eval-warning into the users terminal.

> Keep the alias for at least one full release cycle so that users on stable have time to migrate, then convert it to a `throw`.

Reviewer: A point that has been raised in the past is that for a *fully* staged deprecation cycle in the presence of third‐party packages and modules that try to support both stable and unstable, you actually need the alternative to be available and working in the oldest supported release before you start emitting a warning on unstable. Otherwise, that third‐party code needs to do version detection to avoid emitting warnings for its users on the oldest supported stable release. That stretches it out even longer, though, which exacerbates the problems with manual toil.

> Keep the alias for at least one full release cycle so that users on stable have time to migrate, then convert it to a `throw`.

Reviewer 3: does leaving it for one release cycle mean "at least 6 months" or does it also permit "add a warning 2026-05-01 before the release and convert it to a throw 2026-06-01 after the release"?

Reviewer 2: > does it also permit "add a warning 2026-05-01 before the release and convert it to a throw 2026-06-01 after the release"?

Almost, but the wording was careful to say "release cycle" not just "release".

A "release cycle" is more than just the moment a release branches off, it's the entire period from master rolling onto a new unstable release to that release being, well, released.

Currently, master is "in" the 26.11 release cycle.

So if a warning was added in 26.05 (even if seconds before branch-off), you'd have to wait for 26.11's "full cycle" to finish before converting to `throw`. I.e., immediately after 26.11's release, when master rolls onto 27.05 unstable (at the earliest).

(the fact this needed clarification hints that the wording could be improved further, though)

```nix
{
  # 26.05: warn and redirect
  jbidwatcher = warnAlias "'jbidwatcher' has been renamed to 'jbidwatcher2'" jbidwatcher2; # Added 2026-05-01

  # 26.11 (or later): convert to throw
  jbidwatcher = throw "'jbidwatcher' has been removed. Use 'jbidwatcher2' instead."; # Converted to throw 2026-11-01
}
```

> **When no replacement exists**, consider adding a `lib.warn` to the package expression for one release before removing it, so users get a heads-up while their system still builds.

Reviewer: This won’t work, per my main review comment.

Author: Probably needs `meta.problems` first agreed. Should we still document the intent?

What if we want to remove a whole package set? like `agdaPackages`  Would we just turn it into a `throw` from one day to another? what is the expected path today? 

I want to do two things:

- document status quo
- guidance for maintainers to prevent downstream frustration (when avoidable)

Reviewer 2: In #480338 I abused `meta.knownVulnerabilities` to warn users that a package is unmaintained upstream, and therefore implicitly/potentially vulnerable.

The advice as-is may still be valid if it was re-phrased as _"move your package to `aliases.nix` and add a `lib.warnOnInstantiate`"_ (aside, I think this is preferred to plain `lib.warn`?), because packages only declared via `aliases.nix` aren't evaluated by CI.

That lack of CI coverage is not ideal either, though, so maybe we don't want to encourage that pattern.

We could mention something like:

> if the reason for deprecating relates to a known vulnerabilities, including lack of upstream maintenance, you could consider making users aware via `meta.knownVulnerabilities`.

as an interim solution until `meta.problems` supports package deprecation?

**When immediate removal is necessary** (e.g., security vulnerability, package already fails to build), go directly to a `throw`.

For bigger removals (an entire package set, a widely-used library, a language ecosystem), the **warning phase** matters even more.

> Post in the [Breaking changes thread on Discourse](https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574) *before* you merge the removal.

Reviewer: There’s already a delay between changes getting merged and reaching the channels. I don’t think notifying people before they can actually adjust their configurations is a good idea in most cases.

Reviewer 2: Depends on the audience. If you're notifying end-users (which `Breaking changes thread on Discourse` is) I agree.

But I could see value in trying to get wider feedback and awareness from contributors, via the relevant matrix rooms and discorse topics. Whether we want to require that extra step is another question, though. Personally, I'd probably forget to do it most of the time.

### Writing good messages

The `throw` message is the only thing users see when their evaluation breaks.
We don't know how good everyone is at finding github PRs and issues. To show users we care about them.
Tell them what happened and what to do about it:

```nix
# "[What happened]. [Optional migration instruction]"
# ->
throw

>   "'pkgs.foolers.foo' has been removed. Use 'foo' from the top-level package set instead (pkgs.foo)." # Added 2026-03-29

Reviewer: This example would actually break the script, because it can’t handle multi‐line aliases 🫠

Author: Yeah nixfmt greets. I had this as a single line but ci failed

```

> Always include a date comment (`# Added YYYY-MM-DD`) so that stale throws can be identified and pruned by `maintainers/scripts/remove-old-aliases.py`.

Reviewer: FWIW, part of the problem with the script is that it’s based on dates (of commit authorship, not even merge), which don’t correspond cleanly to releases, so it can actually transition things to the next stage prematurely. That’s one of the things we were hoping to address with the work on aliases in `by-name`, but there are technical obstacles to solve.

1. Create a new branch for your change, e.g. `git checkout -b jbidwatcher`
1. Remove the actual package including its directory, e.g. `git rm -rf pkgs/applications/misc/jbidwatcher`
1. Remove the package from the list of all packages (`pkgs/top-level/all-packages.nix`).
1. Add an alias in `pkgs/top-level/aliases.nix` (There is also `pkgs/applications/editors/vim/plugins/aliases.nix`.
   Package sets typically do not have aliases, so we can't add them there.)

    If the package has been moved or renamed, use `warnAlias` (see above).
    If it has been removed entirely, use a `throw` with a good message:

    ```nix
    {
      jbidwatcher = throw "'jbidwatcher' has been removed because the project was discontinued and the ebay login it depends on no longer works."; # Added 2021-03-15
    }
    ```

2. Test if the changes introduced any issues by running `nix-env -qaP -f . --show-trace`.
   It should show the list of packages without errors.
3. Commit the changes.
   Explain why the package was removed.

>    If it was declared discontinued upstream, add a link to the source.

Reviewer: I don’t think we need to encourage people to put things in commit messages if they don’t add anything over what’s already explained in the diff. (Existing problem, though.)

Reviewer 2: I think this depends. I see commit messages as serving two purposes:
1. give a motivation for the changes, or explain intent
   - diffs rarely speak for themselves in terms of _why_ something was done
2. summarize large changes
   - this is _by definition_ repeating what's in the diff
   - but it's only relevant for _large_ diffs, where someone might prefer to understand the changes by reading the commit message instead of its patches
